### PR TITLE
[broker] Fix topic produced through REST not support Authorization

### PR DESCRIPTION
**Pull request by liudezhi2098**
Master Issue: #13766
## Motivation
fix topic produced through REST not support Authorization
## Documentation
- [x] `no-need-doc`

### CODE REVIEW

**Comment by mattisonchao** [comment]
Hi , @liudezhi2098 

Could you add some test for this change ?

**Comment by thomasechen** [comment]
Hi @codelipenghui , The fix is critical to the existing community users who want to use the REST function. I suggest to add the fix as part of the release 2.9.2.

**Comment by thomasechen** [comment]
> Hi @codelipenghui , The fix is critical to the existing community users who want to use the REST function. I suggest to add the fix as part of the release 2.9.2.

Sorry , I mean the users that already implemented the Pulsar in their environment

**Comment by codelipenghui** [comment]
@thomasechen Yes, we can contain it in 2.9.2

**Review by zymap — APPROVED** [review]

**Review by MarvinCai — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicsAuthTest.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+                + Base64.getEncoder().encodeToString(SECRET_KEY.getEncoded()));
+        Set<String> superUserRoles = new HashSet<>();
+        superUserRoles.add("admin");
+        conf.setSuperUserRoles(superUserRoles);
+        Set<String> providers = new HashSet<>();
+        providers.add(AuthenticationProviderToken.class.getName());
+        conf.setAuthenticationProviders(providers);
+        super.internalSetup();
+        topics = spy(new Topics());
+        topics.setPulsar(pulsar);
```
Comment (Demogorgon314) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicsAuthTest.java:96]: topics never used, we can remove it.

**Review by Demogorgon314 — COMMENTED** [review]
Overall LGTM, just left a few comments about code style.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicsAuthTest.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                {CONSUME_TOKEN, 401},
+                {PRODUCE_TOKEN, 200}
+        };
+    }
+
+    @Test(dataProvider = "variations")
+    public void testProduceToNonPartitionedTopic(String token, int status) throws Exception {
+        admin.topics().createNonPartitionedTopic("persistent://" + testTenant + "/"
+                + testNamespace + "/" + testTopicName);
+        AsyncResponse asyncResponse = PowerMockito.mock(AsyncResponse.class);
```
[the remark this answers is not in the document]
Comment (Demogorgon314) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicsAuthTest.java:131]: Others `asyncResponse ` can be removed either.

**Review by hangc0276 — APPROVED** [review]

**Review by Shoothzj — APPROVED** [review]

**Review by Demogorgon314 — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicsAuthTest.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    @DataProvider(name = "variations")
+    public static Object[][] variations() {
+        return new Object[][]{
+                {CONSUME_TOKEN, 401},
+                {PRODUCE_TOKEN, 200}
+        };
+    }
+
+    @Test(dataProvider = "variations")
+    public void testProduceToNonPartitionedTopic(String token, int status) throws Exception {
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicsAuthTest.java:124]: There are some repeated code blocks, maybe we could try to reuse the same code block.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicsAuthTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        producerMessages.setMessages(ObjectMapperFactory.getThreadLocal().readValue(message,
+                new TypeReference<List<ProducerMessage>>() {
+                }));
+
+        WebTarget root = buildWebClient();
+        Response response = root.path("/topics/persistent/" + testTenant + "/" + testNamespace + "/"
+                + testTopicName).request(MediaType.APPLICATION_JSON)
+                .header("Authorization", "Bearer " + token)
+                .post(Entity.json(producerMessages));
+        Assert.assertEquals(response.getStatus(), status);
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicsAuthTest.java:145]: There are some repeated code blocks, maybe we could try to reuse the same code block. It seems that only topic names are different.

**Review by gaoran10 — APPROVED** [review]

**Comment by thomasechen** [comment]
@liudezhi2098

 I have done the testing about producing the message through REST client. The HTTP could produce the message successfully if I granted the authorization in the namespace or topic level. However, if I use the token which belongs to tenant admin,the system will still show the message under below.  

"**Unauthorized to produce to topic persistent://tls/demo3/hello_thomas with clientAppId [tls] and authdata org.apache.pulsar.broker.authentication.AuthenticationDataHttps@318c6550"}**"

Actually, if we use the client library (Java, Node), we can produce the message to the topic using the tenant admin token  (the tenant admin role was set ) without granting additional permission to namespace and topics.    

**Reproduce :** 

1.  I have a tenant called tls which admin role is also called "tls"
image

2.  There is no permission settings related to the namespace tls/demo4
image

3.  We use the token of the role "tls" to produce a message and succeed
image
image

4. We use "POSTMAN" to produce a message to the same topic
    4.1 Authorization Settings: 
   image
    
    4.2 Message Body Detail
 image


5. The system will give me the unauthorized message
image
  
6. If we grant the permission allowing producing message to the namespace tls/demo4, then we can successfully produce the message
   6.1  Grant the permission to the role tls
  image
  
   6.2  Produce the message successfully
image

  **Conclusion:**
  REST authorization mechanism should be as same as the other clients (JAVA, NODEJS ......)

**Pull request merged**